### ISSUE TITLE
prov/util: have av handle sockaddr_in sin_zero field being non-zero

### ISSUE THREAD

**Issue by soumagne** [ofiwg/libfabric#7450]
those errors started to appear randomly after some recent changes I made to our code (so it could be that I'm now doing something terribly wrong or that the timing of things slightly changed). In any case, it looks like this happens only with prov/tcp and sockets does not show the issue (I have yet to try with verbs rxm). Somehow it looks like either connection or connection after an fi_av_remove() tend to produce those errors (this is with master but it looks like earlier versions also produce that):
```
libfabric:246890:1645052875::ofi_rxm:ep_ctrl:rxm_handle_error():753<warn> fi_eq_readerr: err: Connection refused (111), prov_err: Success (0)
libfabric:246890:1645052875::ofi_rxm:ep_ctrl:rxm_handle_error():753<warn> fi_eq_readerr: err: Connection refused (111), prov_err: Success (0)
libfabric:246890:1645052875::ofi_rxm:ep_ctrl:rxm_handle_error():753<warn> fi_eq_readerr: err: Connection refused (111), prov_err: Success (0)
```
I have no idea at the moment why this happens, it looks like some sort of race in the connection management where I can see that one of peers receives a nack cmd. `fi_av_insert()` is successful and both peers are up and running. I'll debug some more but I was wondering if that rings some bell.

**Comment by shefty** [comment]
Are you seeing any other issues?

This print shouldn't be a warning as the event isn't uncommon.  It usually indicates both peers tried to connect to each other simultaneously, and one of the connections was discarded.  MPI hits this frequently, and it is timing related.

**Comment by soumagne** [comment]
when that happens I'm not able to communicate with the peer, i.e. a send returns EAGAIN and I get those messages over and over.

**Comment by shefty** [comment]
Can you provide more details around what you were attempting to do with the fi_av_remove call?  AFAIK, that's not a heavily used path, so there could be undiscovered opportunities there.

**Comment by soumagne** [comment]
okay, it took me a while to figure that one out but it looks like somehow it is interpreting the `sin_zero` field of `struct sockaddr_in`? (I use FI_SOCKADDR_IN in that case) I had somehow a struct that was not memset to 0 and was only setting the fields for `sin_addr`, `sin_port` and `sin_family` and inserting that into the AV. This is what was causing the issue to appear some/most of the time after my changes.

We don't really use `fi_av_remove()`, we were using it at some point to force removal of a peer that was detected dead. Here it was mostly for testing.

**Comment by soumagne** [comment]
I think it might be good though to have a check on `fi_av_insert()` that ensures that `sin_zero` is set to 0. That would probably help to get an error early on.

**Comment by shefty** [comment]
Renamed issue title to reflect request.  Will need to decide if we want the app to set sin_zero to 0 and fail it not, or have libfabric clear that field.  I'm leaning toward the first option.  I don't think this would break apps, since it looks like things would fail anyway if that field is left unset.  But I need to analyze that in more detail.

**Comment by shefty** [comment]
Fixed by #7473
